neqo-transport/src/cc/cubic.rs

```diff
@@ -0,0 +1,193 @@
+    /// take to increase to W_max. We use bytes not MSS units, therefore this
+    /// equation will be: W_cubic(t) = C*MSS*(t-K)^3 + W_max.
+    ///
+    /// From that equation we can calculate K as:
+    /// K = cubic_root((W_max - W_cubic) / C / MSS);
+    fn calc_k(&self, curr_cwnd: f64) -> f64 {
+        ((self.w_max - curr_cwnd) / CUBIC_C / MAX_DATAGRAM_SIZE_F64).cbrt()
+    }
+
+    /// W_cubic(t) = C*(t-K)^3 + W_max (Eq. 1)
+    /// t is relative to the start of the congestion avoidance phase and it is in seconds.
+    fn w_cubic(&self, t: f64) -> f64 {
+        CUBIC_C * (t - self.k).powi(3) * MAX_DATAGRAM_SIZE_F64 + self.w_max
+    }
+}
+
+impl WindowAdjustment for Cubic {
+    // This is because of the cast in the last line from f64 to usize.
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    fn on_packets_acked(
+        &mut self,
+        curr_cwnd: usize,
+        new_acked_bytes: usize,
+        min_rtt: Duration,
+        now: Instant,
+    ) -> usize {
+        let curr_cwnd_f64 = convert_to_f64(curr_cwnd);
+        self.tcp_acked_bytes += f64::try_from(u32::try_from(new_acked_bytes).unwrap()).unwrap();
+        if self.ca_epoch_start.is_none() {
+            // This is a start of a new congestion avoidance phase.
+            self.ca_epoch_start = Some(now);
+            // reset tcp_acked_bytes and estimated_tcp_cwnd;
+            self.tcp_acked_bytes = f64::try_from(u32::try_from(new_acked_bytes).unwrap()).unwrap();
+            self.estimated_tcp_cwnd = curr_cwnd_f64;
+            if self.last_max_cwnd <= curr_cwnd_f64 {
+                self.w_max = curr_cwnd_f64;
+                self.k = 0.0;
+            } else {
+                self.w_max = self.last_max_cwnd;
+                self.k = self.calc_k(curr_cwnd_f64);
+            }
+            qtrace!([self], "New epoch");
+        }
+
+        let time_ca = self
+            .ca_epoch_start
+            .map_or(min_rtt, |t| now + min_rtt - t)
+            .as_secs_f64();
+        let target = self.w_cubic(time_ca);
+
+        let mut cnt = if target > curr_cwnd_f64 {
+            curr_cwnd_f64 / (target - curr_cwnd_f64) * MAX_DATAGRAM_SIZE_F64
+        } else {
+            100.0 * curr_cwnd_f64
+        };
+
+        let tcp_cnt = self.estimated_tcp_cwnd / CUBIC_ALPHA;
+        while self.tcp_acked_bytes > tcp_cnt {
+            self.tcp_acked_bytes -= tcp_cnt;
+            self.estimated_tcp_cwnd += MAX_DATAGRAM_SIZE_F64;
+        }
+
+        if self.estimated_tcp_cwnd > curr_cwnd_f64 && self.estimated_tcp_cwnd > target {
+            let cnt_tcp_equation =
+                curr_cwnd_f64 / (self.estimated_tcp_cwnd - curr_cwnd_f64) * MAX_DATAGRAM_SIZE_F64;
+            if cnt > cnt_tcp_equation {
+                cnt = cnt_tcp_equation;
+            }
+        }
+
+        // Limit increase to max 1 MSS per 2 ack packets.
+        cnt = cnt.max(2.0 * MAX_DATAGRAM_SIZE_F64);
```

Comment:
I missed this last time around, but it might be worth pointing out that this limit is a direct result of the limit of 1.5cwnd in the target calculation.  You can draw a straight line from that to this, but it's non-obvious.  Also, I'm not sure that 2 is necessarily the only value we need to consider here, so it needs a constant.

BTW, this changes the function into an exponential function.  The increase goes from linear (the TCP region), to cubic (the W_max calculation), to exponential.  It's not really accurate to call it cubic, other than to note that a cubic function is used to smooth the transition from linear to exponential.